src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Testing/Common.cs

```diff
@@ -13,12 +19,58 @@ static Common()
             TestTracer = new Tracer(settings);
             ServiceName = TestTracer.DefaultServiceName;
 
+            Tracer.Instance = TestTracer;
+
             // Preload environment variables.
             CIEnvironmentValues.DecorateSpan(null);
         }
 
         internal static Tracer TestTracer { get; private set; }
 
         internal static string ServiceName { get; private set; }
+
+        internal static void FlushSpans(IntegrationInfo integrationInfo)
+        {
+            if (TestTracer.Settings.IsIntegrationEnabled(integrationInfo))
+            {
+                FlushSpans();
+            }
+        }
+
+        internal static void FlushSpans()
+        {
+            try
+            {
+                // https://docs.microsoft.com/en-us/archive/msdn-magazine/2015/july/async-programming-brownfield-async-development#the-thread-pool-hack
+                var flushTask = Task.Run(() => InternalFlush());
+                flushTask.GetAwaiter().GetResult();
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Exception occurred when flushing spans.");
+            }
+
+            static async Task InternalFlush()
+            {
+                try
+                {
+                    // We have to ensure the flush of the buffer after we finish the tests of an assembly.
+                    // For some reason, sometimes when all test are finished none of the callbacks to handling the tracer disposal is triggered.
+                    // So the last spans in buffer aren't send to the agent.
+                    Log.Debug("Integration flushing spans.");
+                    await TestTracer.FlushAsync().ConfigureAwait(false);
+                    // The current agent writer FlushAsync method can return inmediately if a payload is being sent (there is buffer lock)
+                    // There is not api in the agent writer that guarantees the send has been sucessfully completed.
+                    // Until we change the behavior of the agentwriter we should at least wait 2 seconds before returning.
+                    Log.Debug("Waiting 2 seconds to flush.");
+                    await Task.Delay(2000).ConfigureAwait(false);
```

Comment:
This is only flushing on _assembly_ unload right? Not for individual tests/theorys? 2s seems like a reasonable trade off at the assembly level, but not at the test level!

Also though, 2s is pretty arbitrary, is it worth us adding an API to the agent writer to e.g. only return after all buffers are flushed? Or alternatively, if we're not actually closing the Tracer, do we _have_ to make sure everything's flushed at this point?

Reply:
> This is only flushing on assembly unload right? Not for individual tests/theorys? 2s seems like a reasonable trade off at the assembly level, but not at the test level!

Yes this for assembly unload only.

> Also though, 2s is pretty arbitrary, is it worth us adding an API to the agent writer to e.g. only return after all buffers are flushed? Or alternatively, if we're not actually closing the Tracer, do we have to make sure everything's flushed at this point?

I talked with @kevingosse a little bit about supporting this return after flush on the agent writer, this also will fix some scenarios where `ProcessExit` and `CancelKeyPress` events are involved. In some cases we don't close the tracer and only call for flushing, but we don't have guarantees that buffers are actually flushed after that call, and the process will close after.

We need to make sure the buffer is flushed at this point because I found that the testhost can finish without rising neither the `ProcessExit` and `CancelKeyPress` events.